libzecale/circuits/pairing/bls12_377_membership_check_gadgets.tcc

```diff
@@ -0,0 +1,205 @@
+
+    // Untwist:
+    //   untwist_x =
+    //     (x as Fp6) * bls12_377_g2_untwist_frobenius_twist_v.coeffs[0]
+    //   untwist_y =
+    //     (y as Fp12) * bls12_377_g2_untwist_frobenius_twist_w_3
+    Fp6_3over2_variable<Fq6T> x_as_Fp6(
+        pb,
+        *g2.X,
+        libsnark::Fp2_variable<FqeT>(
+            pb, libff::Fqe<nppT>::zero(), FMT(annotation_prefix, " Fqe(0)")),
+        libsnark::Fp2_variable<FqeT>(
+            pb, libff::Fqe<nppT>::zero(), FMT(annotation_prefix, " Fqe(0)")),
+        FMT(annotation_prefix, " x_as_Fp6"));
+    Fp6_3over2_variable<Fq6T> y_as_Fp6(
+        pb,
+        *g2.Y,
+        libsnark::Fp2_variable<FqeT>(
+            pb, FqeT::zero(), FMT(annotation_prefix, " Fqe(0)")),
+        libsnark::Fp2_variable<FqeT>(
+            pb, FqeT::zero(), FMT(annotation_prefix, " Fqe(0)")),
+        FMT(annotation_prefix, " y_as_Fp6"));
+    Fp12_2over3over2_variable<FqkT> y_as_Fp12(
+        pb,
+        y_as_Fp6,
+        Fp6_3over2_variable<Fq6T>(
+            pb, Fq6T::zero(), FMT(annotation_prefix, " Fp6(0)")),
+        FMT(annotation_prefix, " y_as_Fp12"));
+
+    Fp6_3over2_variable<Fq6T> untwist_x =
+        x_as_Fp6 * libff::bls12_377_g2_untwist_frobenius_twist_v.coeffs[0];
+    Fp12_2over3over2_variable<FqkT> untwist_y =
+        y_as_Fp12 * libff::bls12_377_g2_untwist_frobenius_twist_w_3;
+
+    // Frobenius:
+    Fp6_3over2_variable<Fq6T> untwist_frob_x = untwist_x.frobenius_map(exp);
+    Fp12_2over3over2_variable<FqkT> untwist_frob_y =
+        untwist_y.frobenius_map(exp);
+
+    // Twist:
+    Fp6_3over2_variable<Fq6T> untwist_frob_twist_x =
+        untwist_frob_x *
+        libff::bls12_377_g2_untwist_frobenius_twist_v_inverse.coeffs[0];
+    Fp12_2over3over2_variable<FqkT> untwist_frob_twist_y =
+        untwist_frob_y *
+        libff::bls12_377_g2_untwist_frobenius_twist_w_3_inverse;
+
+    return libsnark::G2_variable<wppT>(
+        pb,
+        untwist_frob_twist_x._c0,
+        untwist_frob_twist_y._c0._c0,
+        annotation_prefix);
+}
+
+// bls12_377_G2_membership_check_gadget
+
+template<typename wppT>
+bls12_377_G2_membership_check_gadget<wppT>::
+    bls12_377_G2_membership_check_gadget(
+        libsnark::protoboard<libff::Fr<wppT>> &pb,
+        libsnark::G2_variable<wppT> &P,
+        const std::string &annotation_prefix)
+    : libsnark::gadget<libff::Fr<wppT>>(pb, annotation_prefix)
+    , _P_checker(pb, P, FMT(annotation_prefix, " _P_checker"))
+    // \psi(P) - P
+    , _psi_P_minus_P(
+          pb,
+          bls12_377_g2_untwist_frobenius_twist(
+              pb, P, 1, FMT(annotation_prefix, " psi(P)")),
+          g2_variable_negate(pb, P, FMT(annotation_prefix, " -P")),
+          libsnark::G2_variable<wppT>(pb, FMT(annotation_prefix, " psi(P)-P")),
+          FMT(annotation_prefix, " _psi_P_minus_P"))
+    // [t](\psi(P) - P)
+    , _t_times_psi_P_minus_P(
+          pb,
+          libff::bls12_377_trace_of_frobenius,
+          _psi_P_minus_P._C,
```

Comment:
I was originally going to us `result`, but thought it may be better to be consistent with the equivalent G1 gadgets in libsnark.  We have to have an inconsistency somewhere, so I thought better to have all EC add and dbl gadgets be similar, but we could rename.   Worth nothing that we have code that is generic over these G1/G2 gadgets, although that code doesn't actually actually use the members explicitly.

Reply:
Thanks @dtebbs I see. I think I would favor `result` as it makes the code easier to grasp in my opinion. However if all similar gadgets use the A+B=C pattern, then we can stick to that, it's not a major one (just forces the reader to inspect the gadget to "click" and remember that C is the "result wire").
What about adding a comment:
```c++
_psi_P_minus_P._C, // result (or "output") of _psi_P_minus_P
```
?

Reply:
Renamed member.

Reply:
Great, thanks